src/Tizen.System.Session/Session/Session.cs

```diff
@@ -0,0 +1,509 @@
+                }
+            }
+        }
+
+        private void OnSwitchUserCompletion(SubsessionEventInfoNative infoNative, IntPtr data)
+        {
+            _switchUserCompletionHandler?.Invoke(this, new SwitchUserCompletionEventArgs(infoNative));
+        }
+
+        private Interop.Session.SubsessionEventCallback _switchUserCompletionCB = null;
+
+        private event EventHandler<SwitchUserCompletionEventArgs> _switchUserCompletionHandler = null;
+
+        private readonly object _switchUserCompletionLock = new object();
+
+        /// <summary>
+        /// Event to be invoked when an existing subession user is successfully switched to an acvite state.
+        /// </summary>
+        /// <exception cref="ArgumentException">Session UID of this object is invalid</exception>
+        /// <exception cref="OutOfMemoryException">Out of memory</exception>
+        /// <exception cref="IOException">Internal error</exception>
+        /// <exception cref="UnauthorizedAccessException">Not permitted</exception>
+        /// <exception cref="NotSupportedException">Not supported</exception>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public event EventHandler<SwitchUserCompletionEventArgs> SwitchUserCompletion
+        {
+            add
+            {
+                lock (_switchUserCompletionLock)
+                {
+                    if (_switchUserCompletionHandler == null)
+                        RegisterCallbackForEvent(SessionEventType.SwitchUserCompletion, _switchUserCompletionCB, OnSwitchUserCompletion);
+                    _switchUserCompletionHandler += value;
+                }
+            }
+            remove
+            {
+                lock (_switchUserCompletionLock)
+                {
+                    _switchUserCompletionHandler -= value;
+                    if (_switchUserCompletionHandler == null)
+                        UnregisterCallbackForEvent(SessionEventType.SwitchUserCompletion, _switchUserCompletionCB);
+                }
+            }
+        }
+
+        private void CheckError(SessionError ret, string msg)
+        {
+            if (ret == SessionError.None)
+                return;
+
+            Log.Error(SessionErrorFactory.LogTag, msg);
+            SessionErrorFactory.ThrowException(ret);
+        }
+
+        private void IntPtrToStringArray(IntPtr unmanagedArray, int size, out string[] managedArray)
+        {
+            managedArray = new string[size];
+
+            byte[] byteArray = new byte[MaxUserLength * size];
+            byte[] tmpArray = new byte[MaxUserLength];
+
+            Marshal.Copy(unmanagedArray, byteArray, 0, MaxUserLength * size);
+
+            for (int i = 0; i < size; i++)
+            {
+                Array.Copy(byteArray, MaxUserLength * i, tmpArray, 0, MaxUserLength);
+                managedArray[i] = Encoding.UTF8.GetString(tmpArray);
+            }
+        }
```

Comment:
It seems to be wrong
Please explain your intent

What is native type of `unmanagedArray`. We usually don't copy native buffer to get string

Reply:
This may be a bit misleading, I admit. The original array definition looks like this:

    typedef char subsession_user_t[SUBSESSION_USER_MAXLEN];
    ...
    EXPORT_API int subsession_get_user_list(int session_uid, subsession_user_t **user_list, int *user_count)
    {
        ....
        subsession_user_t *const list = calloc(elems, sizeof *list);
        ....
        *user_list = list;
    }
    
which means it's actually an array of fixed size (SUBSESSION_USER_MAXLEN) arrays. Unfortunately, a simple IntPtrToStringArray implementation:

        static void IntPtrToStringArray(IntPtr unmanagedArray, int size, out string[] managedArray)
        {
            managedArray = new string[size];
            IntPtr[] IntPtrArray = new IntPtr[size];

            Marshal.Copy(unmanagedArray, IntPtrArray, 0, size);

            for (int iterator = 0; iterator < size; iterator++)
            {
                managedArray[iterator] = Marshal.PtrToStringAnsi(IntPtrArray[iterator]);
            }
        }

doesn't work in this case.

Reply:
so, unmanagedArray is sequence of char and length is 20 * elm_no
in this case, 
You can marshal like it
``` c#
    static void IntPtrToStringArray(IntPtr unmanagedArray, int size, out string[] managedArray)
    {
        managedArray = new string[size]; 
       var curr = unmanagedArray;
        for (int iterator = 0; iterator < size; iterator++)
        {
            managedArray[iterator] = Marshal.PtrToStringAnsi(curr, 20);
            curr = IntPtr.Add(curr, 20);
        }
    }

```

I didn't test this code, it could be contained error carefully check it

Reply:
Thanks for suggestion. Will investigate it further and check it.

Reply:
We tested it yesterday and it works. We'll include it in our changes. Thank you.